Sources/AsyncHTTPClient/RequestBag+StateMachine.swift

```diff
@@ -0,0 +1,529 @@
+        case .executing(let executor, let requestState, .buffering(var buffer, next: .askExecutorForMore)):
+            self.state = .modifying
+
+            if let byteBuffer = buffer.popFirst() {
+                self.state = .executing(executor, requestState, .buffering(buffer, next: .askExecutorForMore))
+                return .consume(byteBuffer)
+            }
+
+            // buffer is empty, wait for more
+            self.state = .executing(executor, requestState, .waitingForRemote(buffer))
+            return .requestMoreFromExecutor(executor)
+
+        case .executing(let executor, let requestState, .buffering(var buffer, next: .eof)):
+            self.state = .modifying
+
+            if let byteBuffer = buffer.popFirst() {
+                self.state = .executing(executor, requestState, .buffering(buffer, next: .eof))
+                return .consume(byteBuffer)
+            }
+
+            // buffer is empty, wait for more
+            self.state = .finished(error: nil)
+            return .finishStream
+
+        case .executing(_, _, .buffering(_, next: .error(let error))):
+            self.state = .finished(error: error)
+            return .failTask(error, executorToCancel: nil)
+
+        case .executing(_, _, .waitingForRemote):
+            preconditionFailure("Invalid state... We just returned from a consumption function. We can't already be waiting")
+
+        case .redirected:
+            return .doNothing
+
+        case .finished(error: .some):
+            return .doNothing
+
+        case .finished(error: .none):
+            preconditionFailure("Invalid state... If no error occured, this must not be called, after the request was finished")
+
+        case .modifying:
+            preconditionFailure()
+        }
+    }
+
+    enum FailAction {
+        case failTask(HTTPRequestScheduler?, HTTPRequestExecutor?)
+        case cancelExecutor(HTTPRequestExecutor)
+        case none
+    }
+
+    mutating func fail(_ error: Error) -> FailAction {
+        switch self.state {
+        case .initialized:
+            self.state = .finished(error: error)
+            return .failTask(nil, nil)
+        case .queued(let queuer):
+            self.state = .finished(error: error)
+            return .failTask(queuer, nil)
+        case .executing(let executor, let requestState, .buffering(_, next: .eof)):
+            self.state = .executing(executor, requestState, .buffering(.init(), next: .error(error)))
+            return .cancelExecutor(executor)
+        case .executing(let executor, let requestState, .buffering(_, next: .askExecutorForMore)):
+            self.state = .executing(executor, requestState, .buffering(.init(), next: .error(error)))
+            return .cancelExecutor(executor)
+        case .executing(let executor, _, .buffering(_, next: .error(_))):
+            // this would override another error, let's keep the first one
+            return .cancelExecutor(executor)
```

Comment:
If `next:` is `.error` already, might we have already cancelled the executor? And does that matter? Does executors support being cancelled multiple times?

Reply:
The executor supports being cancelled a bazillion times :)